### PR TITLE
Generalizing reward models

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#1005 

A hopefully quick add in parallel to my PPO PR to generalize reward models to also support e.g. llama2 reward models. 
Model builders for the previous so-called classifier models are now correctly named as reward models, since they only classify a single label. I've also moved checkpoint conversion logic to `modules/rlhf`. This checkpointing logic can generally be useful for any classifier model, but I've tried to keep things simple and contained here without leaking the scope of this functionality to the rest of the API.

Previously, if someone wanted to use classifiers for their own recipes, they'd use the underlying component builder and set the model type to `MISTRAL_REWARD`. This hasn't really changed - they just now use `REWARD`, and the model builders are a bit clearer to indicate their default use is for reward modelling.

#### Test plan

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/modules/rlhf/utils/_convert_weights.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def mistral_reward_hf_to_tune(
 ) -> Dict[str, torch.Tensor]:
     """
     Convert a state dict from HF's format to torchtune's format, which contains the weights
-    of a Mistral reward model.
+    of a reward model (i.e. a classifier with a single class).
```
Comment (felipemello1) [inline torchtune/modules/rlhf/utils/_convert_weights.py:38]: I guess this confirms that classifier is a case of a reward, and maybe we should swap the current naming?

**Review by felipemello1 — COMMENTED** [review]
Great PR! Thanks for writing it.

I have some questions, a few of them are simple, but others may be more towards the core maintainers.

File: torchtune/models/llama2/_model_builders.py [hunk 2]
```diff
@@ -263,9 +264,91 @@ def lora_llama2_70b(
...
 that LoRA is applied to are quantized per the QLoRA paper: https://arxiv.org/abs/2305.14314.
 Please see `lora_llama2_70b` for full API arguments.
 """
+
+
+def llama2_reward_7b() -> TransformerDecoder:
+    """
+    Builder for creating a Llama2 model initialized w/ the default 7B parameter values
+    from https://arxiv.org/abs/2307.09288, where the output layer is a classification layer
+    projecting to a single class for reward modelling.
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/models/llama2/_model_builders.py:280]: There's actually a typo here, sorry, `llama2_classifier_7b` shouldn't exist. I think we're on the same page. The general pattern here is: `_classifier` exists in `_component_builders` as a general interfacew to a classifier model. The only use for these models currently in the codebase is for reward modelling for PPO, where reward models output a scalar (I do agree this constraint doesn't apply to all reward models). 

For simplicity, I could drop the `_reward` in model builders, and expose a `num_classes` in the model builder which is defaulted to 1. So you'd have:

```python
model_family/model_builders.py

... import model_family_classifier

def model_family_7b_classifier(num_classes: int = 1):
    return model_family_classifier(num_classes=num_classes, ...)

```

I think this would clean up the API, and offer flexibility to users who wish to define recipes for classification models  for which there is some interest from the community e.g. https://github.com/pytorch/torchtune/issues/1124.

How does this sound?

**Comment by SalmanMohammadi** [comment]
Thanks so much for the review @felipemello1. Apologies for not providing a bit more context about the PR to help make things clearer. I discussed this briefly with @pbontrager on Discord in the "Discussing PPO" thread (which has been my defacto channel for pinging devs for contributing questions - maybe I should move to #dev..), it'd be great you have your input on some of the points we raised.

As you've correctly gathered, the main use for these classifier models is to support my work on RLHF to allow the use of pre-trained reward models during PPO. My primary aim for this PR was to reduce the review overhead on #1005, and I'd like my changes to be minimal and non-invasive (in that they can easily be refactored), and rely on future refactors (such as https://github.com/pytorch/torchtune/issues/1017#issuecomment-2155725874 ) to address some of the potential long-term effects on the codebase. 

> Are we comfortable with having 9 builders per model size? Maybe the answer is yes, and thats ok

There's been a few suggestions around this (see some comments in the original PR https://github.com/pytorch/torchtune/pull/837) - generally, there's some agreement that we can probably define these classifiers in a more minimal way like you suggested. However, we're touching on support for general classifiers vs support for classifiers as they're needed now in the codebase here. For now, I'd argue that we should revisit this further down the line when the need for generalisation is apparent and/or in a `TransformerDecoder` refactor.

File: torchtune/models/llama2/_model_builders.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/models/llama2/_model_builders.py:280]: I think the way you have it here where classifier is a generic component builder and reward is just num_classes = 1 in the model builders is great. We also don't want to add too many builder layers that add distance between what the user calls and the actual model architecture definition

File: torchtune/modules/rlhf/utils/_convert_weights.py [hunk 3]
```diff
@@ -76,17 +81,17 @@ def _permute(t, n_heads):
     return converted_state_dict
 
 
-def mistral_reward_tune_to_hf(
+def reward_tune_to_hf(
```
Comment (RdoubleA) [inline torchtune/modules/rlhf/utils/_convert_weights.py:84]: I wonder if there's a better location for these - our other converters are in models/convert_weights. PEFT adapter conversions were added there. I think it makes sense to also throw these in there, or do you think that will make the file bloated?

Comment (SalmanMohammadi) [reply torchtune/modules/rlhf/utils/_convert_weights.py:84]: Yeah I was thinking similar. I think it does make the file a bit bloating and would reflect the sentiment from my comment about; we could revisit this and put weight conversion somewhere more general when we need it.

File: torchtune/models/llama2/_model_builders.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/models/llama2/_model_builders.py:280]: Thanks @RdoubleA. Let's roll with things as-is for now.

**Comment by felipemello1** [comment]
> For now, I'd argue that we should revisit this further down the line

sounds good to me! Specially since you are already aligned with Philip and Rafi. Thanks for your reply!

**Review by RdoubleA — COMMENTED** [review]
Overall looks good to me. Do we have tests for the reward convert weights utilities?

**Comment by SalmanMohammadi** [comment]
> Overall looks good to me. Do we have tests for the reward convert weights utilities?

There's a test here https://github.com/pytorch/torchtune/blob/570dc8f829dfaf2adf5a186ab1b93e11fb925bb4/tests/torchtune/utils/test_checkpointer.py#L443

**Review by RdoubleA — APPROVED** [review]
